pkg/hive/hive.go

```diff
@@ -260,57 +272,102 @@ func (s *Service) checkAndAddPeers(ctx context.Context, peers pb.Peers) {
 
 	var peersToAdd []swarm.Address
 	mtx := sync.Mutex{}
-	wg := sync.WaitGroup{}
 
-	for _, p := range peers.Peers {
-		err := s.sem.Acquire(ctx, 1)
-		if err != nil {
-			return
-		}
+	if s.bootnode {
+		for _, p := range peers.Peers {
+			multiUnderlay, err := ma.NewMultiaddrBytes(p.Underlay)
+			if err != nil {
+				s.logger.Errorf("hive: multi address underlay err: %v", err)
+				continue
+			}
 
-		wg.Add(1)
-		go func(newPeer *pb.BzzAddress) {
-			defer func() {
-				s.sem.Release(1)
-				wg.Done()
-			}()
+			bzzAddress := bzz.Address{
+				Overlay:     swarm.NewAddress(p.Overlay),
+				Underlay:    multiUnderlay,
+				Signature:   p.Signature,
+				Transaction: p.Transaction,
+			}
 
-			multiUnderlay, err := ma.NewMultiaddrBytes(newPeer.Underlay)
+			err = s.addressBook.Put(bzzAddress.Overlay, bzzAddress)
 			if err != nil {
-				s.logger.Errorf("hive: multi address underlay err: %v", err)
-				return
+				s.logger.Warningf("skipping peer in response %s: %v", p.String(), err)
+				continue
 			}
 
-			ctx, cancel := context.WithTimeout(ctx, pingTimeout)
-			defer cancel()
+			peersToAdd = append(peersToAdd, bzzAddress.Overlay)
+		}
+	} else {
 
-			// check if the underlay is usable by doing a raw ping using libp2p
-			if _, err = s.streamer.Ping(ctx, multiUnderlay); err != nil {
-				s.metrics.UnreachablePeers.Inc()
-				s.logger.Debugf("hive: peer %s: underlay %s not reachable", hex.EncodeToString(newPeer.Overlay), multiUnderlay)
-				return
+		wg := sync.WaitGroup{}
+		for _, p := range peers.Peers {
+
+			overlay := swarm.NewAddress(p.Overlay)
+			cacheOverlay := overlay.ByteString()[:cachePrefix]
+
+			// cached peer, skip
+			if _, ok := s.lru.Get(cacheOverlay); ok {
+				continue
 			}
 
-			bzzAddress := bzz.Address{
-				Overlay:     swarm.NewAddress(newPeer.Overlay),
-				Underlay:    multiUnderlay,
-				Signature:   newPeer.Signature,
-				Transaction: newPeer.Transaction,
+			// mark peer as seen
+			_ = s.lru.Add(cacheOverlay, nil)
+
+			// if peer exists already in the addressBook, skip
+			_, err := s.addressBook.Get(overlay)
+			if err == nil {
+				continue
+			} else {
+				s.logger.Errorf("hive: addressbook %v", err)
```

Comment:
this means that you return an error here if the peer is _not_ in the address book, so you basically fail for every peer you don't know

Reply:
@mrekucci's request was that we should at least log it, but I agree, logging here is does not add much value.